cassandane/Cassandane/Cyrus/SearchFuzzy.pm

```diff
@@ -1926,5 +1920,306 @@ sub test_reindex_mb_uniqueid
+    ) || die;
+    $self->make_message("msg2",
+        mime_type => "multipart/related",
+        mime_boundary => "123456789abcdef",
+        body => ""
+        ."\r\n--123456789abcdef\r\n"
+        ."Content-Type: text/plain\r\n"
+        ."\r\n"
+        ."bodyterm"
+        ."\r\n--123456789abcdef\r\n"
+        ."Content-Type: application/pdf\r\n"
+        ."\r\n"
+        ."attachterm"
+        ."\r\n--123456789abcdef--\r\n"
+    ) || die;
+
+    xlog "Run squatter with cachedir $cachedir";
+    $self->{instance}->run_command({cyrus => 1},
+        'squatter', "--attachextract-cache-dir=$cachedir", @squatterArgs);
+}
+
+sub test_squatter_attachextract_cache
+    :min_version_3_9 :needs_search_xapian :SearchAttachmentExtractor
+{
+    my ($self) = @_;
+    my $instance = $self->{instance};
+    my $imap = $self->{store}->get_client();
+
+    my $tracedir = tempdir(DIR => $instance->{basedir} . "/tmp");
+    $self->start_echo_extractor(tracedir => $tracedir);
+
+    xlog "Create and index index messages";
+    my $cachedir = tempdir(DIR => $instance->{basedir} . "/tmp");
+    $self->squatter_attachextract_cache_run($cachedir);
+
+    xlog "Assert text bodies of both messages are indexed";
+    my $uids = $imap->search('fuzzy', 'body', 'bodyterm');
+    $self->assert_deep_equals([1,2], $uids);
+
+    xlog "Assert attachments of both messages are indexed";
+    $uids = $imap->search('fuzzy', 'xattachmentbody', 'attachterm');
+    $self->assert_deep_equals([1,2], $uids);
+
+    xlog "Assert extractor only got called once";
+    my @tracefiles = glob($tracedir."/*_PUT_*");
+    $self->assert_num_equals(1, scalar @tracefiles);
+
+    xlog "Assert cache contains one file";
+    my @files = glob($cachedir."/*");
+    $self->assert_num_equals(1, scalar @files);
+}
+
+sub test_squatter_attachextract_cachedir_noperm
+    :min_version_3_9 :needs_search_xapian :SearchAttachmentExtractor :NoCheckSyslog
+{
+    my ($self) = @_;
+    my $instance = $self->{instance};
+    my $imap = $self->{store}->get_client();
+
+    my $tracedir = tempdir(DIR => $instance->{basedir} . "/tmp");
+    $self->start_echo_extractor(tracedir => $tracedir);
+
+    xlog "Run squatter with read-only cache directory";
+    my $cachedir = tempdir(DIR => $instance->{basedir} . "/tmp");
+    chmod "400", $cachedir;
+    $self->squatter_attachextract_cache_run($cachedir, "--allow-partials");
+
+    xlog "Assert text bodies of both messages are indexed";
+    my $uids = $imap->search('fuzzy', 'body', 'bodyterm');
+    $self->assert_deep_equals([1,2], $uids);
+
+    xlog "Assert attachments of both messages are not indexed";
+    $uids = $imap->search('fuzzy', 'xattachmentbody', 'attachterm');
+    $self->assert_deep_equals([], $uids);
+
+    xlog "Assert extractor got called twice with attachment uploads";
+    my @tracefiles = glob($tracedir."/*_PUT_*");
+    $self->assert_num_equals(2, scalar @tracefiles);
+
+    xlog "Assert cache contains no file";
+    chmod "0700", $cachedir;
```

Comment:
Surprisingly, `"0700"` has made no difference: the cachedir still winds up left with the same weird permissions that cleanup can't remove later.  I've added some debug logging locally and it's definitely this test and this line breaking it (and not the earlier `chmod "400"`, with or without a leading zero) but I haven't figured out why yet...

Here's the logging output around the 400 change:

```
[1851468] =====> Cyrus::SearchFuzzy::test_squatter_attachextract_cachedir_noperm[2057] XXX setting 400 on /dev/shm/cass/0123540101/tmp/wPQhdRo1Oq
[1851468] =====> Cyrus::SearchFuzzy::test_squatter_attachextract_cachedir_noperm[2058] XXX permissions before: drwx------ 2 cyrus cyrus 40 Apr 21 11:23 /dev/shm/cass/0123540101/tmp/wPQhdRo1Oq

[1851468] =====> Cyrus::SearchFuzzy::test_squatter_attachextract_cachedir_noperm[2060] XXX permissions after: drw--w---- 2 cyrus cyrus 40 Apr 21 11:23 /dev/shm/cass/0123540101/tmp/wPQhdRo1Oq
````

And then here's the logging output around the 0700 change:

```
[1851468] =====> Cyrus::SearchFuzzy::test_squatter_attachextract_cachedir_noperm[2076] XXX setting 0700 on /dev/shm/cass/0123540101/tmp/wPQhdRo1Oq
[1851468] =====> Cyrus::SearchFuzzy::test_squatter_attachextract_cachedir_noperm[2077] XXX permissions before: drw--w---- 2 cyrus cyrus 40 Apr 21 11:23 /dev/shm/cass/0123540101/tmp/wPQhdRo1Oq

[1851468] =====> Cyrus::SearchFuzzy::test_squatter_attachextract_cachedir_noperm[2079] XXX permissions after: d-w-rwxr-T 2 cyrus cyrus 40 Apr 21 11:23 /dev/shm/cass/0123540101/tmp/wPQhdRo1Oq
```

What the hell is going on here?

Here's the patch with my logging -- if you add this patch and run it yourself with -vvv -fpretty, what do you see?  The XXX lines will be coloured yellow/orange in the output.

```
diff --git a/cassandane/Cassandane/Cyrus/SearchFuzzy.pm b/cassandane/Cassandane/Cyrus/Sea>
index 9211dfd2c..ad4829548 100644
--- a/cassandane/Cassandane/Cyrus/SearchFuzzy.pm
+++ b/cassandane/Cassandane/Cyrus/SearchFuzzy.pm
@@ -2054,7 +2054,10 @@ sub test_squatter_attachextract_cachedir_noperm
 
     xlog "Run squatter with read-only cache directory";
     my $cachedir = tempdir(DIR => $instance->{basedir} . "/tmp");
-    chmod "400", $cachedir;
+    xlog "XXX setting 400 on $cachedir";
+    xlog "XXX permissions before: " . qx{ls -ld $cachedir};
+    chmod("400", $cachedir) || die;
+    xlog "XXX permissions after: " . qx{ls -ld $cachedir};
     $self->squatter_attachextract_cache_run($cachedir, "--allow-partials");
 
     xlog "Assert text bodies of both messages are indexed";
@@ -2070,7 +2073,10 @@ sub test_squatter_attachextract_cachedir_noperm
     $self->assert_num_equals(2, scalar @tracefiles);
 
     xlog "Assert cache contains no file";
-    chmod "0700", $cachedir;
+    xlog "XXX setting 0700 on $cachedir";
+    xlog "XXX permissions before: " . qx{ls -ld $cachedir};
+    chmod("0700", $cachedir) || die;
+    xlog "XXX permissions after: " . qx{ls -ld $cachedir};
     my @files = glob($cachedir."/*");
     $self->assert_num_equals(0, scalar @files);
 }
```

Reply:
Did you forget to push your changes?  I went to re-review just now, but the PR still has `chmod "400"` and `chmod "0700"`

Reply:
So sorry about that, pushed now.